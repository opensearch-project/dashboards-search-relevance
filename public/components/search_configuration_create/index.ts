/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { SearchConfigurationCreateWithRouter } from './search_configuration_create';
export { ResultsPanel } from './components/results_panel';
export { ValidationPanel } from './components/validation_panel';
export { SearchConfigurationForm } from './components/search_configuration_form';
