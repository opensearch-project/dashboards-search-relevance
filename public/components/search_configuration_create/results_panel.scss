.resultsPanel__scrollContainer {
  width: 100%;
  overflow-x: auto;
  max-width: 100%;
}

.osd-table {
  min-width: max-content;
} 