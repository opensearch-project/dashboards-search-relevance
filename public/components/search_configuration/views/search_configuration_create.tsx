/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiPageHeader,
  EuiPageTemplate,
  EuiPanel,
  EuiFlexGroup,
  EuiFlexItem,
} from '@elastic/eui';
import React, { useCallback } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { CoreStart, NotificationsStart } from '../../../../../../src/core/public';
import { SearchConfigurationForm } from '../components/search_configuration_form';
import { ValidationPanel } from '../components/validation_panel';
import { useSearchConfigurationForm } from '../hooks/use_search_configuration_form';

interface SearchConfigurationCreateProps extends RouteComponentProps {
  http: CoreStart['http'];
  notifications: NotificationsStart;
}

export const SearchConfigurationCreate: React.FC<SearchConfigurationCreateProps> = ({
  http,
  notifications,
  history,
}) => {
  const {
    // Form state
    name,
    setName,
    nameError,
    validateNameField,
    query,
    setQuery,
    queryError,
    setQueryError,
    searchTemplate,
    setSearchTemplate,

    // Index state
    indexOptions,
    selectedIndex,
    setSelectedIndex,
    isLoadingIndexes,

    // Pipeline state
    pipelineOptions,
    selectedPipeline,
    setSelectedPipeline,
    isLoadingPipelines,

    // Validation state
    testSearchText,
    setTestSearchText,
    isValidating,
    searchResults,

    // Actions
    validateSearchQuery,
    createSearchConfiguration,
  } = useSearchConfigurationForm({
    http,
    notifications,
    onSuccess: () => history.push('/searchConfiguration'),
  });

  // Handle cancel action
  const handleCancel = useCallback(() => {
    history.push('/searchConfiguration');
  }, [history]);

  return (
    <EuiPageTemplate paddingSize="l" restrictWidth="100%">
      <EuiPageHeader
        pageTitle="Search Configuration"
        description="Configure a new search configuration that represents all the aspects of an algorithm."
        rightSideItems={[
          <EuiButtonEmpty
            onClick={handleCancel}
            iconType="cross"
            size="s"
            data-test-subj="cancelSearchConfigurationButton"
          >
            Cancel
          </EuiButtonEmpty>,
          <EuiButton
            onClick={createSearchConfiguration}
            fill
            size="s"
            iconType="check"
            data-test-subj="createSearchConfigurationButton"
            color="primary"
          >
            Create Search Configuration
          </EuiButton>,
        ]}
      />
      <EuiFlexGroup>
        <EuiFlexItem grow={7}>
          <EuiPanel hasBorder={true}>
            <SearchConfigurationForm
              name={name}
              setName={setName}
              nameError={nameError}
              validateName={validateNameField}
              query={query}
              setQuery={setQuery}
              queryError={queryError}
              setQueryError={setQueryError}
              searchTemplate={searchTemplate}
              setSearchTemplate={setSearchTemplate}
              indexOptions={indexOptions}
              selectedIndex={selectedIndex}
              setSelectedIndex={setSelectedIndex}
              isLoadingIndexes={isLoadingIndexes}
              pipelineOptions={pipelineOptions}
              selectedPipeline={selectedPipeline}
              setSelectedPipeline={setSelectedPipeline}
              isLoadingPipelines={isLoadingPipelines}
            />
          </EuiPanel>
        </EuiFlexItem>
        <EuiFlexItem grow={5} style={{ maxWidth: '41.67%', maxHeight: '80vh', overflow: 'auto' }}>
          <ValidationPanel
            testSearchText={testSearchText}
            setTestSearchText={setTestSearchText}
            isValidating={isValidating}
            searchResults={searchResults}
            onValidate={validateSearchQuery}
          />
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiPageTemplate>
  );
};

export const SearchConfigurationCreateWithRouter = withRouter(SearchConfigurationCreate);
