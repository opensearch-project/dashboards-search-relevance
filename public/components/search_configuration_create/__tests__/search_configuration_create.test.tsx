/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { SearchConfigurationCreate } from '../search_configuration_create';

// Mock the service and hooks
jest.mock('../services/search_configuration_service');
jest.mock('../hooks/use_search_configuration_form');
jest.mock('../components/search_configuration_form', () => ({
  SearchConfigurationForm: ({ name, setName }: any) => (
    <div data-test-subj="search-configuration-form">
      <input
        data-test-subj="name-input"
        value={name}
        onChange={(e) => setName(e.target.value)}
      />
    </div>
  ),
}));
jest.mock('../components/validation_panel', () => ({
  ValidationPanel: ({ testSearchText }: any) => (
    <div data-test-subj="validation-panel">{testSearchText}</div>
  ),
}));

const mockFormState = {
  name: 'Test Search Configuration',
  setName: jest.fn(),
  nameError: '',
  validateNameField: jest.fn(),
  query: '{"query": {"match_all": {}}}',
  setQuery: jest.fn(),
  queryError: '',
  setQueryError: jest.fn(),
  searchTemplate: '',
  setSearchTemplate: jest.fn(),
  indexOptions: [{ label: 'test-index', value: 'uuid-1' }],
  selectedIndex: [{ label: 'test-index', value: 'uuid-1' }],
  setSelectedIndex: jest.fn(),
  isLoadingIndexes: false,
  pipelineOptions: [{ label: 'test-pipeline' }],
  selectedPipeline: [],
  setSelectedPipeline: jest.fn(),
  isLoadingPipelines: false,
  testSearchText: 'test query',
  setTestSearchText: jest.fn(),
  isValidating: false,
  searchResults: null,
  validateSearchQuery: jest.fn(),
  createSearchConfiguration: jest.fn(),
};

describe('SearchConfigurationCreate', () => {
  let mockHttp: any;
  let mockNotifications: any;
  let history: any;

  beforeEach(() => {
    mockHttp = {};
    mockNotifications = {
      toasts: {
        addSuccess: jest.fn(),
        addError: jest.fn(),
        addWarning: jest.fn(),
      },
    };
    history = createMemoryHistory();

    // Reset mocks
    jest.clearAllMocks();

    // Mock the hook return value
    (require('../hooks/use_search_configuration_form').useSearchConfigurationForm as jest.Mock).mockReturnValue(
      mockFormState
    );
  });

  const renderComponent = () => {
    return render(
      <Router history={history}>
        <SearchConfigurationCreate
          http={mockHttp}
          notifications={mockNotifications}
          history={history}
          location={history.location}
          match={{ params: {}, isExact: true, path: '', url: '' }}
        />
      </Router>
    );
  };

  it('should render the component correctly', () => {
    renderComponent();

    expect(screen.getByText('Search Configuration')).toBeInTheDocument();
    expect(screen.getByText(/Configure a new search configuration/)).toBeInTheDocument();
    expect(screen.getByTestId('search-configuration-form')).toBeInTheDocument();
    expect(screen.getByTestId('validation-panel')).toBeInTheDocument();
  });

  it('should handle successful form submission', async () => {
    mockFormState.createSearchConfiguration.mockResolvedValue({ success: true });

    renderComponent();

    const createButton = screen.getByTestId('createSearchConfigurationButton');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(mockFormState.createSearchConfiguration).toHaveBeenCalled();
    });
  });

  it('should handle cancel action', () => {
    renderComponent();

    const cancelButton = screen.getByTestId('cancelSearchConfigurationButton');
    fireEvent.click(cancelButton);

    expect(history.location.pathname).toBe('/searchConfiguration');
  });

  it('should display form and validation panel', () => {
    renderComponent();

    expect(screen.getByTestId('search-configuration-form')).toBeInTheDocument();
    expect(screen.getByTestId('validation-panel')).toBeInTheDocument();
    expect(screen.getByText('test query')).toBeInTheDocument();
  });

  it('should render action buttons', () => {
    renderComponent();

    expect(screen.getByTestId('createSearchConfigurationButton')).toBeInTheDocument();
    expect(screen.getByTestId('cancelSearchConfigurationButton')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });
});