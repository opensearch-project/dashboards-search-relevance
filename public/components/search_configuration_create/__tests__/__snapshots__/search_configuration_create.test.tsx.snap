// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchConfigurationCreate should match snapshot 1`] = `
<div
  class="euiPage euiPage--grow euiPageTemplate"
  style="min-height: 460px;"
>
  <div
    class="euiPageBody euiPageBody--borderRadiusNone"
  >
    <div
      class="euiPanel euiPanel--borderRadiusNone euiPanel--plain euiPanel--noShadow euiPanel--noBorder euiPageContent euiPageContent--borderRadiusNone"
      role="main"
    >
      <div
        class="euiPageContentBody euiPage--paddingLarge euiPage--restrictWidth-custom"
        style="max-width: 100%;"
      >
        <header
          class="euiPageHeader euiPageHeader--responsive euiPageHeader--center"
        >
          <div
            class="euiPageHeaderContent"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive euiPageHeaderContent__top"
            >
              <div
                class="euiFlexItem"
              >
                <h1
                  class="euiTitle euiTitle--large"
                >
                  Search Configuration
                </h1>
                <div
                  class="euiSpacer euiSpacer--l"
                />
                <div
                  class="euiText euiText--medium euiText--constrainedWidth"
                >
                  <p>
                    Configure a new search configuration that represents all the aspects of an algorithm.
                  </p>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--wrap euiPageHeaderContent__rightSideItems"
                >
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--small"
                      data-test-subj="cancelSearchConfigurationButton"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonEmpty__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiIcon--inherit euiButtonContent__icon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M7.293 8 3.146 3.854a.5.5 0 1 1 .708-.708L8 7.293l4.146-4.147a.5.5 0 0 1 .708.708L8.707 8l4.147 4.146a.5.5 0 0 1-.708.708L8 8.707l-4.146 4.147a.5.5 0 0 1-.708-.708L7.293 8Z"
                          />
                        </svg>
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Cancel
                        </span>
                      </span>
                    </button>
                  </div>
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <button
                      class="euiButton euiButton--primary euiButton--small euiButton--fill"
                      data-test-subj="createSearchConfigurationButton"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButton__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiIcon--inherit euiButtonContent__icon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M6.5 12a.502.502 0 0 1-.354-.146l-4-4a.502.502 0 0 1 .708-.708L6.5 10.793l6.646-6.647a.502.502 0 0 1 .708.708l-7 7A.502.502 0 0 1 6.5 12"
                            fill-rule="evenodd"
                          />
                        </svg>
                        <span
                          class="euiButton__text"
                        >
                          Create Search Configuration
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </header>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrow7"
          >
            <div
              class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder"
            >
              <div
                data-test-subj="search-configuration-form"
              >
                <input
                  data-test-subj="name-input"
                  value="Test Search Configuration"
                />
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrow5"
            style="max-width: 41.67%; max-height: 80vh; overflow: auto;"
          >
            <div
              data-test-subj="validation-panel"
            >
              test query
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
